### ISSUE TITLE
stop_gradient inconsistencies with lax.scan and lax.while_loop versus explicit computations

### ISSUE THREAD

**Issue by AOKullberg** [google/jax#13710]
### Description

I'm experiencing inconsistencies with the results of stop_gradient for particular combinations of lax operations.
Particularly, I am looking to repeat a particular calculation _i_ times but only calculate the gradient for the last iteration.
I've implemented such an operation below in three distinct ways:
1. A lax.scan operation wrapped in stop_gradient
2. A lax.while_loop wrapped in stop_gradient
3. Explicit calculation of all steps wrapped in stop_gradient

As you can see below, only the explicit version actually returns a non-zero gradient whereas both the scan and while_loop versions seem to stop the gradient flow for all successive operations as well. 

```python
import jax
import jax.numpy as jnp
from jax.scipy.stats import multivariate_normal
max_iter = 5
R = jnp.array([[1.]])
def scan(y, x, theta):
    def body(carry, _):
        x, theta = carry
        x = theta*x
        return (x, theta), _
    (x, theta), _ = jax.lax.stop_gradient(jax.lax.scan(body, (x, theta), jnp.arange(max_iter)))
    x = theta*x
    return multivariate_normal.logpdf(y, mean=theta*x, cov=R), x

def laxwhile(y, x, theta):
    def cond(carry):
        x, theta, i = carry
        return i < max_iter
    def body(carry):
        x, theta, i = carry
        x = theta*x
        i += 1
        return (x, theta, i)
    x, theta, i = jax.lax.stop_gradient(jax.lax.while_loop(cond, body, (x, theta, 0)))
    x = theta*x
    return multivariate_normal.logpdf(y, mean=theta*x, cov=R), x

def explicit(y, x, theta):    
    x = jax.lax.stop_gradient(theta*x)
    x = jax.lax.stop_gradient(theta*x)
    x = jax.lax.stop_gradient(theta*x)
    x = jax.lax.stop_gradient(theta*x)
    x = jax.lax.stop_gradient(theta*x)
    x = theta*x
    return multivariate_normal.logpdf(y, mean=theta*x, cov=R), x

y = jnp.array([1.])
x = jnp.array([0.1])
theta = 1.25
print(jax.value_and_grad(scan, argnums=-1, has_aux=True)(y, x, theta))
print(jax.value_and_grad(laxwhile, argnums=-1, has_aux=True)(y, x, theta))
print(jax.value_and_grad(explicit, argnums=-1, has_aux=True)(y, x, theta))
# ((DeviceArray(-1.0557882, dtype=float32), DeviceArray([0.38146973], dtype=float32)), DeviceArray(0., dtype=float32, weak_type=True))
# ((DeviceArray(-1.0557882, dtype=float32), DeviceArray([0.38146973], dtype=float32)), DeviceArray(0., dtype=float32, weak_type=True))
# ((DeviceArray(-1.0557882, dtype=float32), DeviceArray([0.38146973], dtype=float32)), DeviceArray(0.39914158, dtype=float32, weak_type=True))
```

### What jax/jaxlib version are you using?

jax v0.3.25, jaxlib v0.3.25

### Which accelerator(s) are you using?

CPU

### Additional system info

Python 3.9.7,  Ubuntu 20.04.3 - 5.15.0-56

**Comment by AOKullberg** [comment]
> Your unrolled version doesn't match the `scan` etc. version exactly. Should have been
> 
> ```python
> def explicit(y, x, theta):
>   for i in range(max_iter):
>     theta, x = lax.stop_gradient((theta, theta * x))  # <<< Notice theta is also wrapped here
>   x = theta * x
>   return multivariate_normal.logpdf(y, mean=theta * x, cov=R), x
> ```

I see. Is this expected behavior?
I would still expect gradients to be computed for the two last operations in all of these functions, which is not the displayed behavior.

**Comment by soraros** [comment]
I think it's the expected behaviour, as everything up to `x = theta * x` (the two loop constructions) are wrapped in `lax.stop_gradient`.

If you want the other way around, you could write

```python
def scan(y, x, theta):
  def body(carry, _):
    x, theta = carry
    x = lax.stop_gradient(x * theta)  # <<< Moved to inside of `lax.scan`
    return (x, theta), None
  (x, theta), _ = lax.scan(body, (x, theta), None, length=max_iter)
  x = theta * x
  return multivariate_normal.logpdf(y, mean=theta * x, cov=R), x
```

**Comment by AOKullberg** [comment]
> I think it's the expected behaviour, as everything up to `x = theta * x` (the two loop constructions) are wrapped in `lax.stop_gradient`.
> 
> If you want the other way around, you could write
> 
> ```python
> def scan(y, x, theta):
>   def body(carry, _):
>     x, theta = carry
>     x = lax.stop_gradient(x * theta)  # <<< Moved to inside of `lax.scan`
>     return (x, theta), None
>   (x, theta), _ = lax.scan(body, (x, theta), None, length=max_iter)
>   x = theta * x
>   return multivariate_normal.logpdf(y, mean=theta * x, cov=R), x
> ```

That seems reasonable, thanks for the clarification.
I realized that I was unnecessarily carrying theta through the whiles/scans anyway, i.e., the resulting implementation
```python
def scan(y, x, theta):
    def body(carry, _):
        x = carry
        x = theta*x
        return x, _
    x, _ = jax.lax.stop_gradient(jax.lax.scan(body, x, jnp.arange(max_iter)))
    x = theta*x
    return multivariate_normal.logpdf(y, mean=theta*x, cov=R), x
```
also works as expected.
Thanks for the help @soraros.
